Crash on launch & background task after battery optimisations

<!-- READ THIS FIRST:
- Make sure you run the latest version of the Android app
- Make sure you run the latest version of Home Assistant
- Make sure to check the Companion docs for troubleshooting and configuration: https://companion.home-assistant.io/
- Make sure the bug you found is not already reported, we love to put work in bugfixes instead of closing duplicate bug reports
  DO NOT DELETE ANY TEXT from this template! All requested information is important.
-->

<!-- If you are reporting an issue for Wear OS please list both devices for the below 3 fields -->
**Home Assistant Android app version(s):**

**Android version(s):** 12

**Device model(s):** Samsung Galaxy S10e

**Home Assistant version:** 2024.6.1-full

**Last working Home Assistant release (if known):** 2024.6.1-full

**Description of problem, include YAML if issue is related to notifications:**

Yesterday I got a notification from the battery optimiser that I hadn't used the Home Assistant companion app for a while (which is not true, I actively use it a few times a week, but ignore that) and that it would be put to sleep in the background. I stupidly allowed this to happen even though I do rely on it as a background service. The battery optimiser also removes all permissions from the app with the expectation that it'll ask for them again on the next use.
Now, the application crashes every 15 minutes in the background, and also instantly on launch. 

I use mTLS authentication for accessing my instance and I believe the battery optimisation has also removed access to the mTLS certificate I have registered to it from the system.

I tried clearing the app cache storage but that didn't help. Even clearing the full app data didn't help. It'd ask me a few times to enter the server details and select the mTLS certificate, and after a few tries would get straight back into crashing again. I had to fully uninstall and reinstall the app to get it working again, so there may be something in the system that has revoked the proper access to the certificate.

Here is the callstack of the crash from the logcat logs.
```
D CompatibilityChangeReporter: Compat change id reported: 160794467; UID 10370; state: ENABLED
D KeyChain: isKeyChainUri: home assistant
D OpenGLRenderer: RenderThread::requireGlContext()
E AndroidRuntime: FATAL EXCEPTION: DefaultDispatcher-worker-3
E AndroidRuntime: Process: io.homeassistant.companion.android, PID: 16883
E AndroidRuntime: java.lang.AssertionError: could not resolve KeyChainService
E AndroidRuntime:        at android.security.KeyChain.bindAsUser(KeyChain.java:1165)
E AndroidRuntime:        at android.security.KeyChain.bindAsUser(KeyChain.java:1070)
E AndroidRuntime:        at android.security.KeyChain.bind(KeyChain.java:1056)
E AndroidRuntime:        at android.security.KeyChain.getCertificateChain(KeyChain.java:909)
E AndroidRuntime:        at io.homeassistant.companion.android.common.data.keychain.KeyChainRepositoryImpl.doLoad(KeyChainRepositoryImpl.kt:65)
E AndroidRuntime:        at io.homeassistant.companion.android.common.data.keychain.KeyChainRepositoryImpl.access$doLoad(KeyChainRepositoryImpl.kt:14)
E AndroidRuntime:        at io.homeassistant.companion.android.common.data.keychain.KeyChainRepositoryImpl$load$3.invokeSuspend(KeyChainRepositoryImpl.kt:41)
E AndroidRuntime:        at io.homeassistant.companion.android.common.data.keychain.KeyChainRepositoryImpl$load$3.invoke(Unknown Source:8)
E AndroidRuntime:        at io.homeassistant.companion.android.common.data.keychain.KeyChainRepositoryImpl$load$3.invoke(Unknown Source:4)
E AndroidRuntime:        at kotlinx.coroutines.intrinsics.UndispatchedKt.startUndispatchedOrReturn(Undispatched.kt:61)
E AndroidRuntime:        at kotlinx.coroutines.BuildersKt__Builders_commonKt.withContext(Builders.common.kt:163)
E AndroidRuntime:        at kotlinx.coroutines.BuildersKt.withContext(Unknown Source:1)
E AndroidRuntime:        at io.homeassistant.companion.android.common.data.keychain.KeyChainRepositoryImpl.load(KeyChainRepositoryImpl.kt:36)
E AndroidRuntime:        at io.homeassistant.companion.android.HomeAssistantApplication$onCreate$3.invokeSuspend(HomeAssistantApplication.kt:78)
E AndroidRuntime:        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
E AndroidRuntime:        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:104)
E AndroidRuntime:        at kotlinx.coroutines.internal.LimitedDispatcher$Worker.run(LimitedDispatcher.kt:111)
E AndroidRuntime:        at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:99)
E AndroidRuntime:        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:585)
E AndroidRuntime:        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:802)
E AndroidRuntime:        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:706)
E AndroidRuntime:        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:693)
E AndroidRuntime:        Suppressed: kotlinx.coroutines.internal.DiagnosticCoroutineContextException: [StandaloneCoroutine{Cancelling}@2bbf122, Dispatchers.IO]
D Debug   : low && ship && 3rdparty app crash, do not dump
W ActivityManager: crash : io.homeassistant.companion.android,10370
```

As an aside, I don't see the Home Assistant Companion app listed in the list of apps that I can set to be excluded from battery optimisations so I don't know if I can stop this from happening again.

<!--
- For Wear OS devices we will need the LogCat logs from the device.
- For Android Auto the logs can be retrieved from the connected device.
- Logs from the device can be taken from Settings > Companion App Troubleshooting > Show and Share Logs
-->
**Companion App Logs:** Unable to get them since the app was crashing. I had a rummage through `/data/data/io.homeassistant.companion.android` but didn't see anything that looked relevant.

```

```

**Screenshot or video of problem:**

**Additional information:**

--- a/common/src/main/java/io/homeassistant/companion/android/common/data/keychain/KeyChainRepositoryImpl.kt
+++ b/common/src/main/java/io/homeassistant/companion/android/common/data/keychain/KeyChainRepositoryImpl.kt
@@ -63,16 +63,24 @@ class KeyChainRepositoryImpl @Inject constructor(
             if (chain == null) {
                 chain = try {
                     KeyChain.getCertificateChain(context, alias!!)
-                } catch (e: Exception) {
-                    Log.e(TAG, "Exception getting certificate chain", e)
+                } catch (t: Throwable) {
+                    when (t) {
+                        is AssertionError,
+                        is Exception -> Log.e(TAG, "Issue getting certificate chain", t)
+                        else -> throw t
+                    }
                     null
                 }
             }
             if (key == null) {
                 key = try {
                     KeyChain.getPrivateKey(context, alias!!)
-                } catch (e: Exception) {
-                    Log.e(TAG, "Exception getting private key", e)
+                } catch (t: Throwable) {
+                    when (t) {
+                        is AssertionError,
+                        is Exception -> Log.e(TAG, "Issue getting private key", t)
+                        else -> throw t
+                    }
                     null
                 }
             }

